File: Subtract1WithoutArithmeticOperators.c
#include <stdio.h>

int subtractOne(int x){
    int m = 1;  // Biến m dùng để kiểm tra các bit từ phải qua trái
    // Lật tất cả các bit là 0 cho đến khi gặp một bit là 1
    while (!(x & m)) {
        x = x ^ m;  // Lật bit tại vị trí m nếu bit đó là 0
        m <<= 1;    // Dịch m sang trái để kiểm tra bit tiếp theo
    }
    // Khi gặp bit 1 đầu tiên, lật nó thành 0
    x = x ^ m;  // Lật bit 1 đầu tiên từ phải sang trái thành 0

    return x;   // Trả về kết quả
}
int main(){
    int num;
    printf("Nhập số cần trừ 1: ");
    scanf("%d", &num);  // Nhập giá trị vào biến num

    // Gọi hàm subtractOne và in kết quả
    printf("Số sau khi trừ 1 là: %d\n", subtractOne(num)); 

    return 0;
}

File: Subtract1WithoutArithmeticOperators.py
def subtractOne(x):
    # Khởi tạo biến m là 1 (đại diện cho bit thấp nhất)
    m = 1
    # Lặp cho đến khi tìm thấy bit 1 trong x
    while ((x & m) == False):  # Kiểm tra xem bit thấp nhất của x có phải là 0 không
        # Đảo bit của x với m (chuyển từ 0 thành 1)
        x = x ^ m
        # Di chuyển m sang trái một bit, tức là kiểm tra bit tiếp theo
        m = m << 1
    # Khi gặp bit 1, ta đảo bit đó (trừ đi 1 từ x)
    x = x ^ m
    # Trả về giá trị sau khi trừ 1
    return x

if __name__ == '__main__':
    # Yêu cầu người dùng nhập số n
    n = int(input("Nhập số n: "))
    
    # In ra kết quả sau khi trừ 1 từ n
    print(f"Kết quả sau khi trừ 1 từ {n} là: {subtractOne(n)}")
